Support sbt multi client ui

eatkins opened the pull request:
In order to make the `console` task in sbt work with different terminals, it was necessary to make some minor changes to jline2. This project is effectively dead and replaced by jline3 so it makes sense to just apply our changes here and have sbt use the forked version published under the sbt org.

eatkins committed:
Add build.sbt

eatkins committed:
Add setter for terminal factory

In the sbt use case, there are situations in which we want to change the
value of the global terminal programmatically because there is
downstream code that calls Terminal.get() and we want that call to
return the sbt generated terminal.

eatkins committed:
Use System.in rather than FileDescriptor.in

In sbt, we want to use System.in rather than FileDescriptor.in here
because the scala console calls this api and we need the console to read
from System.in rather than FileDescriptor.in. Otherwise the console can
fight with sbt's inputstream and bytes can be lost. It is necessary to
use wrapper classes for the streams because jline will perform its own
wrapping of the default input and output streams (System.in and
System.out). sbt sets System.in and System.out to streams that have
already been wrapped by jline and the double wrapping breaks the i/o on
windows.

eatkins committed:
Add publish settings

This copies much of the build.sbt for the forked sbt ivy project to set
all the parameters necessary to publish this project to maven.

eatkins committed:
Exclude all target directories in .gitignore

eatkins committed:
Add license file

eed3si9n commented on build.sbt:
```diff
@@ -0,0 +1,36 @@
+ * This project uses java 6 doc syntax that isn't parseable by 1.7 which makes it
+ * impossible to generate docs even with sbt 0.13.
+ */
+Compile / doc := file("/dev/null")
```
We'd need doc for Sonatype publishing, so I'd say we should build this fork on JDK 8.

eatkins replied:
I think an empty doc is actually ok but I could at least change this from /dev/null to the normal doc directory. The only way to get doc to work would probably be to delete all of the doc tags or figure out how to use the java6 doc tool but jabba doesn't even have a working java 6 anymore so it's kind of a pain.

eed3si9n replied:
Here's my PR for sbt build with Travis CI running the tests - https://github.com/sbt/jline2/pull/2

eatkins replied:
I was curious about sonatype publishing, so I published into the `com.swoval.jline` org: https://repo1.maven.org/maven2/com/swoval/jline/jline/2.14.7-sbt-8d4caeb87217c3a5f57b6884d28cfffcbe3edc86/. It worked fine with `Compile / doc :=  file("/dev/null")`

eed3si9n commented on project/plugins.sbt:
```diff
@@ -0,0 +1 @@
+addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "1.0.0")
```
We usually use `addSbtPlugin("com.dwijnand" % "sbt-dynver" % "4.0.0")`

eatkins replied:
Oh. I just copied what was in the ivy repo.

eatkins replied:
From what I can tell, sbt-dynver does not work well with this repo because the tags do not match the formatting scheme.

eed3si9n commented:
I wonder if it's worth trying to shade this using http://eed3si9n.com/jarjar-abrams or sbt-assembly like https://github.com/sbt/zinc/pull/799?

eatkins commented:
Shading would defeat the purpose of this. We are deliberately trying to hijack the version of jline that is used at runtime. If we shaded jline, then scala console would use the regular jline version which would break it for the thin client. Jline2 is dead so there are no real binary compatibility risks.

eed3si9n committed:
Add sbt build

eed3si9n committed:
Update the Javadoc so it builds on JDK 8

eed3si9n committed:
Test / parallelExcution := false

eatkins committed:
Merge remote-tracking branch 'eed3si9n/wip/build' into sbt-terminal-support

eatkins commented:
I just noticed #2. I merged in those changes into this PR. What was your motivation for using `val root = project...`? For projects that are not multi-project, I prefer just having everything top level. You seemed to anticipate part of what I prefer about top level with the `sbt-nocomma` plugin. But given that this will never be a multi-project project, I don't really see the point in not just putting everything top level.

eed3si9n commented:
> What was your motivation for using `val root = project...`?

Originally when `build.sbt` was first created, I'm guessing it was meant to be used bare style, separated by blank lines.

However I've been a proponent of `lazy val root = (project in file("."))` because I believe sbt should aim to reduce its vocabulary. Not to the point of Maven, but be more uniform throughout so people can comfortably copy-paste each other's build, and "culturally" the builds come out similar. Less things to learn hopefully makes learning sbt a little easier.

Over the course of years, I've tried to unify documentation towards that (https://www.scala-sbt.org/1.x/docs/Basic-Def.html), and it also related to deprecating `project/Build.scala` despite being popular among the old guards. I've also consciously migrated any builds that I come across to using multi-project style.

eatkins commented:
Before responding, I just want to make clear that I'm not trying to make a big issue out of this specific build file. My life is not affected either way but I'm putting my 2 cents in because if this were just my project I would 100% use single project style and I will continue to do so in the swoval organization. There is an effectively 0% chance of this project ever becoming a multi-project. There is also a good chance that we will only ever publish one version so it really doesn't matter that much how this project is configured as long as we can publish successfully once.

> ...I believe sbt should aim to reduce its vocabulary. Not to the point of Maven, but be more uniform throughout so people can comfortably copy-paste each other's build, and "culturally" the builds come out similar. Less things to learn hopefully makes learning sbt a little easier.

I hear you but we can't remove single project style so it isn't really possible to reduce the vocabulary. I like the single project style for single project builds. I think it's the most user friendly and the best style for learning sbt. Yes, there is some friction when you start adding additional projects and have to switch to multi-project style. Even there, I sort of like having a separate `build.sbt` in each subproject and then the root just looks like:
```
val foo = project
val bar = project
val baz = project
```
The drawback to this is that anything you want to share between subprojects has to be defined in a scala file since nothing can be shared between sbt files.

Adding the `sbt-nocomma` plugin does seem against the goal of reducing the vocabularly. It creates an uncanny valley between the single project style and the multi-project style.

> However I've been a proponent of lazy val root = (project in file("."))

I've also noticed that it's common to use `lazy val` in build.sbt files where regular val is sufficient. My guess is that `lazy val` was introduced in `build.sbt` files where the subprojects were not ordered with dependencies ahead of dependent projects and it was needed to allow forward references. But using `lazy` where it isn't needed is confusing. It would generally be better to order subprojects so that lazy isn't needed to reduce the vocabulary.

eatkins committed:
Switch to multi project style

eatkins committed:
Update license

jline uses the BSD license, not apache v2. My understanding is that
apache is actually slightly more restrictive.

eed3si9n commented:
Yea. All I'm saying is unification is my motivation for always using `lazy val root ...` style. I think there are multiple possible solutions and styles in this space. I'd be happy to discuss more over Hangout and beer (or ginger beer) :)

dwijnand commented:
> I've also noticed that it's common to use `lazy val` in build.sbt files where regular val is sufficient. My guess is that `lazy val` was introduced in `build.sbt` files where the subprojects were not ordered with dependencies ahead of dependent projects and it was needed to allow forward references. But using `lazy` where it isn't needed is confusing. It would generally be better to order subprojects so that lazy isn't needed to reduce the vocabulary.

My preference too, I'm often reordering things to strip them out.  One handy thing you can do is:


```scala
val root = project in file(".")
aggregateProjects(foo, bar)

val foo = project
val bar = project
```

Look, Ma, no `lazy`!  (Because settings are evaluated after vals, or something along those lines.)

eatkins commented:
Cool. Thanks for getting test and doc up and running!
